File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/notifications/Notifier.kt
package com.froggyriia.shiftplanner.notifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.froggyriia.shiftplanner.R

/**
 * Thin wrapper around the system notification service. Delivers real
 * on-device notifications for events the background poller detects; there is no
 * FCM/APNs server push, so this is the "push to phone" path that works without
 * any backend or Firebase project.
 */
object Notifier {

    const val CHANNEL_ID = "shiftplanner_updates"

    fun ensureChannel(context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val manager = context.getSystemService(NotificationManager::class.java) ?: return
            if (manager.getNotificationChannel(CHANNEL_ID) == null) {
                val channel = NotificationChannel(
                    CHANNEL_ID,
                    context.getString(R.string.notif_channel_name),
                    NotificationManager.IMPORTANCE_DEFAULT
                ).apply {
                    description = context.getString(R.string.notif_channel_desc)
                }
                manager.createNotificationChannel(channel)
            }
        }
    }

    fun hasPermission(context: Context): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return true
        return ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.POST_NOTIFICATIONS
        ) == PackageManager.PERMISSION_GRANTED
    }

    /** Posts a notification. No-ops safely if the user hasn't granted permission. */
    fun notify(context: Context, id: Int, title: String, message: String) {
        if (!hasPermission(context)) return
        ensureChannel(context)
        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentTitle(title)
            .setContentText(message)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
            .build()
        runCatching { NotificationManagerCompat.from(context).notify(id, notification) }
    }
}

File: mobile/ios/ShiftPlanner/ShiftPlanner/Shared/Notifications/Notifier.swift
import Foundation
import UserNotifications

/// Thin wrapper around the system notification center. Delivers real
/// on-device notifications for events the poller detects; there is no
/// FCM/APNs server push, so this is the "push to phone" path that works
/// without any backend or push-certificate setup.
enum Notifier {

    /// Ask once for permission. Safe to call repeatedly — the system
    /// remembers the answer and won't re-prompt.
    static func requestPermission() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { _, _ in }
    }

    /// Posts a local notification immediately. No-ops if permission was denied.
    static func notify(id: String, title: String, body: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default

        let request = UNNotificationRequest(identifier: id, content: content, trigger: nil)
        UNUserNotificationCenter.current().add(request)
    }
}
